src/StreamJsonRpc/Reflection/MessageFormatterDuplexPipeTracker.cs

```diff
@@ -0,0 +1,199 @@
+        /// </remarks>
+        /// <exception cref="NotSupportedException">Thrown if no <see cref="MultiplexingStream"/> was provided to the constructor.</exception>
+        public int? GetToken(IDuplexPipe duplexPipe)
+        {
+            Verify.NotDisposed(this);
+
+            MultiplexingStream mxstream = this.GetMultiplexingStreamOrThrow();
+            if (this.RequestIdBeingSerialized is null)
+            {
+                throw new NotSupportedException(Resources.MarshaledObjectInResponseOrNotificationError);
+            }
+
+            if (duplexPipe is null)
+            {
+                return null;
+            }
+
+            MultiplexingStream.Channel channel = mxstream.CreateChannel(new MultiplexingStream.ChannelOptions { ExistingPipe = duplexPipe });
+
+            ImmutableInterlocked.AddOrUpdate(
+                ref this.requestChannelMap,
+                this.RequestIdBeingSerialized.Value,
+                ImmutableList.Create(channel),
+                (key, value) => value.Add(channel));
+
+            // Track open channels to assist in diagnosing abandoned channels.
+            ImmutableInterlocked.TryAdd(ref this.openOutboundChannels, channel.Id, channel);
+            channel.Completion.ContinueWith(_ => ImmutableInterlocked.TryRemove(ref this.openOutboundChannels, channel.Id, out MultiplexingStream.Channel removedChannel), CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.Default).Forget();
+
+            return channel.Id;
+        }
+
+        /// <summary>
+        /// Creates an <see cref="IDuplexPipe"/> from a given token as it is received at the RPC server as a method argument.
+        /// </summary>
+        /// <param name="token">The method argument, which was originally obtained by the client using the <see cref="GetToken(IDuplexPipe)"/> method.</param>
+        /// <returns>The <see cref="IDuplexPipe"/> from the token; or <c>null</c> if <paramref name="token"/> was <c>null</c>.</returns>
+        /// <exception cref="InvalidOperationException">Thrown if the token does not match up with an out of band channel offered by the client.</exception>
+        /// <exception cref="NotSupportedException">Thrown if no <see cref="MultiplexingStream"/> was provided to the constructor.</exception>
+        public IDuplexPipe GetPipe(int? token)
+        {
+            Verify.NotDisposed(this);
+
+            MultiplexingStream mxstream = this.GetMultiplexingStreamOrThrow();
+            if (token is null)
+            {
+                return null;
+            }
+
+            // In the case of multiple overloads, we might be called to convert a channel's token more than once.
+            // But we can only accept the channel once, so look up in a dictionary to see if we've already done this.
+            if (!this.openInboundChannels.TryGetValue(token.Value, out MultiplexingStream.Channel channel))
```

Comment:
In case the request gets rejected with no match, we need to be sure we close this channel rather than leave it open in our dictionary

Reply:
Fixed